pink/barriers/body_spherical_barrier.py

```diff
@@ -0,0 +1,189 @@
+    threshold.
+
+    The barrier function is defined as:
+
+    .. math::
+
+        h(q) = \|p_1(q) -
+        p_2(q)\|^2 - d_{min}^2
+
+    where :math:`p_1(q)` and
+    :math:`p_2(q)` are the positions of the two
+    frames in the world coordinate system, and :math:`d_{min}`
+    is the minimum distance threshold.
+
+    Attributes:
+        frames: Tuple of two frame names.
+        d_min: Minimum distance threshold.
+    """
+
+    frames: Tuple[str, str]
+    d_min: float
+
+    def __init__(
+        self,
+        frames: Tuple[str, str],
+        d_min: float,
+        gain: Union[float, np.ndarray] = 1.0,
+        safe_displacement_gain: float = 3.0,
+    ):
+        """Initialize the BodySphericalBarrier.
+
+        Args:
+            frames: Tuple of two frame names.
+            d_min: Minimum distance threshold.
+            gain: Barrier gain. Defaults to 1.0.
+            safe_displacement_gain: gain for the safe backup displacement
+                cost term. Defaults to 3.0.
+        """
+        if d_min < 0.0:
+            raise NegativeMinimumDistance(
+                "The minimum distance threshold must be non-negative."
+            )
+        super().__init__(
+            dim=1,
+            gain=gain,
+            gain_function=lambda h: h / (1 + np.abs(h)),
+            safe_displacement_gain=safe_displacement_gain,
+        )
+        self.frames = frames
+        self.d_min = d_min
+
+    def compute_barrier(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the value of the barrier function.
+
+        The barrier function is computed based on the Euclidean distance
+        between the two specified frames. It considers the minimum distance
+        threshold.
+
+        The barrier function is given by:
+
+        .. math::
+
+            h(q) = \|p_1(q) -
+            p_2(q)\|^2 - d_{min}^2
+
+        where :math:`p_1(q)` and
+        :math:`p_2(q)` are the positions of the two
+        frames in the world coordinate system, and :math:`d_{min}`
+        is the minimum distance threshold.
+
+        Args:
+            configuration: Robot configuration :math:`q`.
+
+        Returns:
+            Value of the barrier function
+                :math:`h(q)`.
+        """
+        pos1_world, pos2_world = self._get_frame_positions(configuration)
+        return np.array(
+            [np.linalg.norm(pos1_world - pos2_world) ** 2 - self.d_min**2]
+        )
+
+    def compute_jacobian(self, configuration: Configuration) -> np.ndarray:
+        r"""Compute the Jacobian matrix of the barrier function.
+
+        The Jacobian matrix is computed based on the position Jacobians of the
+        two specified frames. The Jacobians are transformed to align with the
+        world coordinate system.
+
+        The Jacobian matrix is given by:
+
+        .. math::
+
+            \frac{\partial h}{\partial q}(q) =
+            2(p_1(q) -
+            p_2(q))^T \begin{bmatrix}
```

Comment:
I don't follow this one: the derivative of

```math
            h(q) = \|p_1(q) - p_2(q)\|^2 - d_{min}^2
```

should be the following, no?

```math
\frac{\partial h}{\partial q} = 2 (p_1 - p_2)^T \left[ \frac{\partial p_1}{\partial q} - \frac{\partial p_2}{\partial q} \right]
```
In that case the Jacobian $\partial h / \partial q$ is the difference rather than the `vstack` of the individual Jacobians $\partial p_i / \partial q$.

Reply:
Yes, you're correct, we've messed something up. The code then simplifies to:

```python
dh_dx = 2 * (pos1_world - pos2_world)
dx_dq = pos1_jac - pos2_jac
return dh_dx.T @ dx_dq
```

I'll also ask @simeon-ned to double-check that the original code was wrong